Mathlib/MeasureTheory/Decomposition/Exhaustion.lean

```diff
@@ -0,0 +1,335 @@
+  If no such set exists (which is only possible if `ν` is not s-finite), we define
+  `μ.sigmaFiniteSetWRT ν = ∅`.
+* `MeasureTheory.Measure.sigmaFiniteSet`: for an s-finite measure `μ`, a measurable set such that
+  `μ.restrict μ.sigmaFiniteSet` is sigma-finite, and for all sets `s ⊆ μ.sigmaFiniteSetᶜ`,
+  either `μ s = 0` or `μ s = ∞`.
+  Defined as `μ.sigmaFiniteSetWRT μ`.
+
+## Main statements
+
+* `measure_eq_top_of_subset_compl_sigmaFiniteSetWRT`: for s-finite `ν`, for all measurable sets `s`
+  in `(sigmaFiniteSetWRT μ ν)ᶜ`, if `ν s ≠ 0` then `μ s = ∞`.
+* An instance showing that `μ.restrict (sigmaFiniteSetWRT μ ν)` is sigma-finite.
+* `restrict_compl_sigmaFiniteSetWRT`: if `μ ≪ ν` and `ν` is s-finite, then
+  `μ.restrict (μ.sigmaFiniteSetWRT ν)ᶜ = ∞ • ν.restrict (μ.sigmaFiniteSetWRT ν)ᶜ`. As a consequence,
+  that restriction is s-finite.
+
+* An instance showing that `μ.restrict μ.sigmaFiniteSet` is sigma-finite.
+* `restrict_compl_sigmaFiniteSet_eq_zero_or_top`: the measure `μ.restrict μ.sigmaFiniteSetᶜ` takes
+  only two values: 0 and ∞ .
+* `measure_compl_sigmaFiniteSet_eq_zero_iff_sigmaFinite`: an s-finite measure `μ` is sigma-finite
+  iff `μ μ.sigmaFiniteSetᶜ = 0`.
+
+## References
+
+* [P. R. Halmos, *Measure theory*, 17.3 and 30.11][halmos1950measure]
+
+-/
+
+open scoped ENNReal Topology
+
+open Filter
+
+namespace MeasureTheory
+
+variable {α : Type*} {mα : MeasurableSpace α} {μ ν : Measure α} {s t : Set α}
+
+open Classical in
+/-- A measurable set such that `μ.restrict (μ.sigmaFiniteSetWRT ν)` is sigma-finite and for all
+measurable sets `t ⊆ sᶜ`, either `ν t = 0` or `μ t = ∞`. -/
+def Measure.sigmaFiniteSetWRT (μ ν : Measure α) : Set α :=
+  if h : ∃ s : Set α, MeasurableSet s ∧ SigmaFinite (μ.restrict s)
+    ∧ (∀ t (_ : MeasurableSet t) (_ : t ⊆ sᶜ) (_ : ν t ≠ 0), μ t = ∞)
+  then h.choose
+  else ∅
+
+@[measurability]
+lemma measurableSet_sigmaFiniteSetWRT :
+    MeasurableSet (μ.sigmaFiniteSetWRT ν) := by
+  rw [Measure.sigmaFiniteSetWRT]
+  split_ifs with h
+  · exact h.choose_spec.1
+  · exact MeasurableSet.empty
+
+instance : SigmaFinite (μ.restrict (μ.sigmaFiniteSetWRT ν)) := by
+  rw [Measure.sigmaFiniteSetWRT]
+  split_ifs with h
+  · exact h.choose_spec.2.1
+  · rw [Measure.restrict_empty]
+    infer_instance
+
+section IsFiniteMeasure
+
+variable [IsFiniteMeasure ν]
+
+/-! We prove that the condition in the definition of `sigmaFiniteSetWRT` is true for finite
+measures. Since every s-finite measure is absolutely continuous with respect to a finite measure,
+the condition will then also be true for s-finite measures. -/
+
+/-- Let `C` be the supremum of `ν s` over all measurable sets `s` such that `μ.restrict s` is
+sigma-finite. `C` is finite since `ν` is a finite measure. Then there exists a measurable set `t`
+with `μ.restrict t` sigma-finite such that `ν t ≥ C - 1/n`. -/
+lemma exists_isSigmaFiniteSet_measure_ge (μ ν : Measure α) [IsFiniteMeasure ν] (n : ℕ) :
```

Comment:
The assumptions `(μ ν : Measure α) [IsFiniteMeasure ν]` show up in this lemma and all the next ones. Can you make them into a `variable` line?

Reply:
I'd rather like to keep all explicit arguments with the lemma definitions.